### PR TITLE
Add logic for gathering/dumping supplies

### PR DESCRIPTION
**Pull request by charliefoxtwo**
Adds `OrderType` for gathering and dumping supplies (they use the same code) and ability to interrupt and resume supply gathering.

### CODE REVIEW

File: src/OpenSage.Game/Logic/Object/Create/CreateModule.cs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 {
     public abstract class CreateModule : BehaviorModule, ICreateModule
     {
-        private bool _shouldCallOnBuildComplete;
+        private bool _shouldCallOnBuildComplete = true;
```
Comment (charliefoxtwo) [inline src/OpenSage.Game/Logic/Object/Create/CreateModule.cs:5]: @tgjones I think you recently modified the code here so I wanted to check with you to see if this has any adverse effects. Currently it's only used in (and is needed for) `SupplyCenterCreate.OnBuildCompleteImpl()`
https://github.com/OpenSAGE/OpenSAGE/blob/71f6625b3b8ee500b9d7626615fa6792c53799fd/src/OpenSage.Game/Logic/Object/Create/SupplyCenterCreate.cs#L14-L20

File: src/OpenSage.Game/Logic/Orders/OrderProcessor.cs [hunk 2]
```diff
@@ -318,7 +318,31 @@ public void Process(IEnumerable<Order> orders)
...
+                            {
+                                continue;
+                            }
+
+                            if (supplyPoint.Definition.KindOf.Get(ObjectKinds.SupplySource))
+                            {
+                                behavior.CurrentSupplySource = supplyPoint;
+                                behavior.SupplyGatherState = SupplyAIUpdate.SupplyGatherStates.SearchingForSupplySource;
+                            }
+                            else // if it's not a supply source, it's a supply center
```
[the remark this answers is not in the document]
Comment (paavohuhtala) [reply src/OpenSage.Game/Logic/Orders/OrderProcessor.cs:339]: I don't undertand how this works. How can you create a `GatherDumpSupplies` order with a supply center selected?

**Review by paavohuhtala — COMMENTED** [review]
Overall code looks very good, just a few questions.

File: src/OpenSage.Game/Logic/Orders/OrderProcessor.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (charliefoxtwo) [reply src/OpenSage.Game/Logic/Orders/OrderProcessor.cs:339]: I believe only supply trucks have `SupplyAIUpdate`. The selected units here are supply trucks (and potentially other units like dozers or combat units which don't have `SupplyAIUpdate`, so we skip them).

`supplyPoint` can be a supply center or a supply source (like a crate or warehouse), and is the object specified in the order.

File: src/OpenSage.Game/Logic/OrderGenerators/UnitOrderGenerator.cs [hunk 3]
```diff
@@ -45,6 +45,13 @@ public string GetCursor(KeyModifiers keyModifiers)
                 // TODO: Should take allies into account.
                 if (_worldObject.Owner != _game.Scene3D.LocalPlayer)
                 {
+                    if (_game.Scene3D.LocalPlayer.SelectedUnits.Any(u => u.Definition.KindOf.Get(ObjectKinds.Harvester)) &&
```
[the remark this answers is not in the document]
Comment (charliefoxtwo) [reply src/OpenSage.Game/Logic/OrderGenerators/UnitOrderGenerator.cs:48]: Looks like one was added back in May, actually! Didn't notice before. I'll change this to use that.

https://github.com/OpenSAGE/OpenSAGE/blob/ad4d613eed0199794c4f05973eb25d7f901ada55/src/OpenSage.Game/Logic/Object/GameObject.cs#L522

[the remark this answers is not in the document]
Comment (charliefoxtwo) [reply src/OpenSage.Game/Logic/OrderGenerators/UnitOrderGenerator.cs:48]: I did a find all for `.Definition.KindOf.Get` and there are 24 other instances across 11 files that we could probably clean up in a single PR.

File: src/OpenSage.Game/Logic/Orders/OrderProcessor.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (paavohuhtala) [reply src/OpenSage.Game/Logic/Orders/OrderProcessor.cs:339]: Ah right, I somehow misread that this was testing the selected unit. 😄 Then it looks good.

**Review by Tarcontar — APPROVED** [review]

**Comment by Tarcontar** [comment]
please update the comment according to @paavohuhtala suggestion and fixe the conflicts. Otherwise it looks good!:)

**Comment by charliefoxtwo** [comment]
> please update the comment according to @paavohuhtala suggestion and fixe the conflicts. Otherwise it looks good!:)

Done!

**Review by Tarcontar — APPROVED** [review]

**Pull request merged**